File: lang/csharp/src/vpkpp/Format/OL.cs
using System;

namespace sourcepp.vpkpp.Format;

using OpenPropertyRequest = Func<PackFile, OpenProperty, byte[]>;

using EntryCallback = Action<string, Entry>;

public class OL : PackFile
{
	protected OL(nint handle, bool managed = true) : base(handle, managed)
	{
	}

	public new static OL? Open(string path, EntryCallback? callback = null, OpenPropertyRequest? _ = null)
	{
		var handle = DLL.vpkpp_ol_open(path, callback is not null ? (entryPath, entry) =>
		{
			callback(entryPath, new Entry(entry, false));
		} : null);
		return handle == nint.Zero ? null : new OL(handle);
	}

	public string Notes
	{
		get
		{
			ThrowIfDisposed();
			return new sourcepp.String(DLL.vpkpp_ol_get_notes(Handle)).Read();
		}
	}

	public string EntryNotes(string path)
	{
		ThrowIfDisposed();
		return new sourcepp.String(DLL.vpkpp_ol_get_entry_notes(Handle, path)).Read();
	}
}

File: lang/c/src/vpkppc/format/OL.cpp
#include <vpkppc/format/OL.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vpkpp;

const char* VPKPP_OL_SIGNATURE = OL_SIGNATURE.data();
const char* VPKPP_OL_EXTENSION = OL_EXTENSION.data();

SOURCEPP_API vpkpp_pack_file_handle_t vpkpp_ol_open(const char* path, vpkpp_entry_callback_t callback) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = OL::open(path, callback ? [callback](const std::string& entryPath, const Entry& entry) {
		callback(entryPath.c_str(), const_cast<Entry*>(&entry));
	} : static_cast<PackFile::EntryCallback>(nullptr));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

SOURCEPP_API sourcepp_string_t vpkpp_ol_get_notes(vpkpp_pack_file_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_STRING_INVALID);

	const auto* ol = dynamic_cast<OL*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN_VAL(ol, SOURCEPP_STRING_INVALID);

	return convert::toString(ol->getNotes());
}

SOURCEPP_API sourcepp_string_t vpkpp_ol_get_entry_notes(vpkpp_pack_file_handle_t handle, const char* path) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_STRING_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(path, SOURCEPP_STRING_INVALID);

	const auto* ol = dynamic_cast<OL*>(convert::handle<PackFile>(handle));
	SOURCEPP_EARLY_RETURN_VAL(ol, SOURCEPP_STRING_INVALID);

	const auto notes = ol->getEntryNotes(path);
	return convert::toString(notes ? *notes : "");
}
